### ISSUE TITLE
Add sugar for kdump configuration

### ISSUE THREAD

**Issue by travier** [coreos/fcct#175]
# What are the configuration options for kdump?

Enabling kdump on FCOS (& RHCOS) requires:
  * setting a kernel argument (for example: `crashkernel=256M`) which has a [default suggested value](https://www.kernel.org/doc/html/latest/admin-guide/kdump/kdump.html?highlight=kdump#boot-into-system-kernel);
  * setting options in `/etc/kdump.conf` to configure where the coredump will be stored;
  * setting options in `/etc/sysconfig/kdump` to configure additional kernel arguments passed to the "dump" kernel instance.

# What would the sugar looks like?

From https://github.com/coreos/fedora-coreos-tracker/issues/622#issuecomment-694225598:

```
kdump:
  crashkernel: 128M
  kargs: foo=bar
  config: |
    net penguin.example.com:/export/cores
```

# References

* [RHEL 8 documentation for examples](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/system_design_guide/installing-and-configuring-kdump_system-design-guide).
* See also initial discussion in https://github.com/coreos/fedora-coreos-tracker/issues/622.

**Comment by bgilbert** [comment]
- Rather than `crashkernel: 128M`, maybe something like `reserved_mib: 128`?
- Is it common to want to pass additional arguments to the dump kernel?  Or can we let the user configure this themselves if they want it?

**Comment by travier** [comment]
> * Rather than `crashkernel: 128M`, maybe something like `reserved_mib: 128`?

The [full option syntax](https://www.kernel.org/doc/html/latest/admin-guide/kdump/kdump.html?highlight=kdump#extended-crashkernel-syntax) enables setting memory size dependent values and offset but I don't know if that's something we want to support.

> * Is it common to want to pass additional arguments to the dump kernel?  Or can we let the user configure this themselves if they want it?

I think this heavily depend on where the coredump will be stored as you might need to setup networking for the crash initrd.

**Comment by bgilbert** [comment]
> The full option syntax enables setting memory size dependent values and offset but I don't know if that's something we want to support.

I suspect that's a pretty unusual setting.  Users can always bypass the sugar and configure advanced settings manually.

> I think this heavily depend on where the coredump will be stored as you might need to setup networking for the crash initrd.

Would it make sense to provide target-specific sugar?

```yaml
kdump:
  reserved_mib: 128
  destinations:
    net:
      kernel_args: ip=X
      remotes:
        - penguin.example.com:/export/cores
    something-else: z

**Comment by kelvinfan001** [comment]
>> - Is it common to want to pass additional arguments to the dump kernel? Or can we let the user configure this themselves if they want it?

> I think this heavily depend on where the coredump will be stored as you might need to setup networking for the crash initrd.

Looking at `kdump.conf` (config for `kdumpctl`) and `[/usr/bin/kdumpctl](https://src.fedoraproject.org/rpms/kexec-tools/blob/master/f/kdumpctl)` (executed by `kdump.service`), it doesn't seem like there is a way to configure the kargs passed to the crash kernel (via `kexec`); it looks like the kargs passed to the crash kernel are always [the same](https://src.fedoraproject.org/rpms/kexec-tools/blob/master/f/kdumpctl#_648). So I think that although `kexec` does support specifying kargs for the "crash/second kernel", it might not be necessary for `kdump` to work.

**Comment by travier** [comment]
It looks like `kdumpctl` sources the content of `/etc/sysconfig/kdump` to setup kdump kernel kargs.

**Comment by travier** [comment]
> I suspect that's a pretty unusual setting. Users can always bypass the sugar and configure advanced settings manually.

Agree.

> > I think this heavily depend on where the coredump will be stored as you might need to setup networking for the crash initrd.
> 
> Would it make sense to provide target-specific sugar?

AFAIK, the [kdump.conf docs](https://src.fedoraproject.org/rpms/kexec-tools/blob/master/f/kdump.conf#_8) say that "only one dump target and path can be specified".

**Comment by kelvinfan001** [comment]
> It looks like kdumpctl sources the content of /etc/sysconfig/kdump to setup kdump kernel kargs.

ahh you're right, I completely missed that. Sorry for the confusion!

**Comment by bgilbert** [comment]
> > Would it make sense to provide target-specific sugar?
>
> AFAIK, the kdump.conf docs say that "only one dump target and path can be specified".

Ah, okay.  Even so, it might be useful to abstract some of the low-level config file details.

**Comment by jlebon** [comment]
> Enabling kdump on FCOS (& RHCOS) requires:

I think that list is also missing enabling `kdump.service` itself.

**Comment by hshiina** [comment]
Hi,

Is this work still required?
Regarding OpenShift, Butane is available for configuring kdump since OpenShift 4.8 as in [documentation](https://docs.openshift.com/container-platform/4.8/support/troubleshooting/troubleshooting-operating-system-issues.html#enabling-kdump-on-day-1). This doesn't look so complicated for those who want to enable kdump simply and looks flexible for those who want to customize the configuration by themselves.

I'm interested in this issue because I hope development of kdump for OpenShift will be completed and it will be fully available.

**Comment by bgilbert** [comment]
It's not required, and we can always add it later.

I'd note, though, that the linked example in OCP docs is not trivial.  It has one service enablement, one karg, and two config files with different syntaxes and a bunch of boilerplate.  This is exactly the sort of thing sugar can help simplify.

**Comment by travier** [comment]
I don't think this feature is required for kdump to be fully supported. However it would definitely help with UX as @bgilbert mentioned.

**Comment by bgilbert** [comment]
Is this something we still want to pursue?

**Comment by travier** [comment]
I'd say that it would be nice to have but it's low priority.

**Comment by yasminvalim** [comment]
Hi! Butane is being merged into the Ignition repository, and the Butane repo will be archived soon.

We opened a new issue to continue this discussion in Ignition: https://github.com/coreos/ignition/issues/2261

Please use that issue for any follow-ups. Thank you!
